pkg/engine/plugins.go

```diff
@@ -0,0 +1,191 @@
+// function. If the language host does not support this operation, the empty set is returned.
+func gatherPluginsFromProgram(plugctx *plugin.Context, prog plugin.ProgInfo) (pluginSet, error) {
+	logging.V(preparePluginLog).Infof("gatherPluginsFromProgram(): gathering plugins from language host")
+	set := newPluginSet()
+	langhostPlugins, err := plugctx.Host.GetRequiredPlugins(prog, plugin.AllPlugins)
+	if err != nil {
+		return set, err
+	}
+	for _, plug := range langhostPlugins {
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromProgram(): plugin %s %s is required by language host", plug.Name, plug.Version)
+		set.Add(plug)
+	}
+	return set, nil
+}
+
+// gatherPluginsFromSnapshot inspects the snapshot associated with the given Target and returns the set of plugins
+// required to operate on the snapshot. The set of plugins is derived from first-class providers saved in the snapshot
+// and the plugins specified in the deployment manifest.
+func gatherPluginsFromSnapshot(plugctx *plugin.Context, target *deploy.Target) (pluginSet, error) {
+	logging.V(preparePluginLog).Infof("gatherPluginsFromSnapshot(): gathering plugins from snapshot")
+	set := newPluginSet()
+	if target == nil || target.Snapshot == nil {
+		logging.V(preparePluginLog).Infof("gatherPluginsFromSnapshot(): no snapshot available, skipping")
+		return set, nil
+	}
+	for _, res := range target.Snapshot.Resources {
+		urn := res.URN
+		if !providers.IsProviderType(urn.Type()) {
+			logging.V(preparePluginVerboseLog).Infof("gatherPluginsFromSnapshot(): skipping %q, not a provider", urn)
+			continue
+		}
+		pkg := providers.GetProviderPackage(urn.Type())
+		version, err := providers.GetProviderVersion(res.Inputs)
+		if err != nil {
+			return set, err
+		}
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromSnapshot(): plugin %s %s is required by first-class provider %q", pkg, version, urn)
+		set.Add(workspace.PluginInfo{
+			Name:    pkg.String(),
+			Kind:    workspace.ResourcePlugin,
+			Version: version,
+		})
+	}
+	for _, plug := range target.Snapshot.Manifest.Plugins {
+		logging.V(preparePluginLog).Infof(
+			"gatherPluginsFromSnapshot(): plugin %s %s is required by snapshot manifest", plug.Name, plug.Version)
+		set.Add(plug)
+	}
+	return set, nil
+}
+
+// ensurePluginsAreInstalled inspects all plugins in the plugin set and, if any plugins are not currently installed,
+// uses the given backend client to install them. Installations are processed in parallel, though
+// ensurePluginsAreInstalled does not return until all installations are completed.
+func ensurePluginsAreInstalled(client deploy.BackendClient, plugins pluginSet) error {
+	if client == nil {
+		logging.V(preparePluginLog).Infoln("ensurePluginsAreInstalled(): skipping due to nil client")
+		return nil
+	}
+	logging.V(preparePluginLog).Infof("ensurePluginsAreInstalled(): beginning")
+	var installTasks errgroup.Group
```

Comment:
From [the docs](https://godoc.org/golang.org/x/sync/errgroup#Group.Go):
> The first call to return a non-nil error cancels the group; its error will be returned by Wait.

Is this the behavior we want, or should we keep trying? I can think of advantages and disadvantages to both.

Reply:
Hmm, I think I'd actually prefer to keep trying, but since this is all best-effort I don't think it's that bad if we bail early if a plugin fails to install. The plan is doomed anyway; the only difference is the state of the machine after `pulumi` exits.